src/journal-remote/journal-gatewayd.c

```diff
@@ -307,42 +345,99 @@ static int request_parse_range(
-                                return -ENOMEM;
+        if (!m->cursor)
+                return -ENOMEM;
 
-                        r = safe_atoi64(t, &m->n_skip);
-                        if (r < 0)
-                                return r;
-                }
+        m->cursor[strcspn(m->cursor, WHITESPACE)] = 0;
+        if (isempty(m->cursor))
+                m->cursor = mfree(m->cursor);
 
-                p = (colon2 ? colon2 : colon) + 1;
-                if (*p) {
-                        r = safe_atou64(p, &m->n_entries);
-                        if (r < 0)
-                                return r;
+        return 0;
+}
 
-                        if (m->n_entries <= 0)
-                                return -EINVAL;
 
-                        m->n_entries_set = true;
-                }
+static int request_parse_range_time(
+                RequestMeta *m,
+                const char *range) {
 
-                m->cursor = strndup(range, colon - range);
+        char *colon, *until;
+        int r;
+
+        range += 5;
+        range += strspn(range, WHITESPACE);
+
+        colon = strchr(range, ':');
+        if (!colon)
+                return -EINVAL;
+
+        if (colon - range > 0) {
+                _cleanup_free_ char *t;
+                t = strndup(range, colon - range);
+                if (!t)
+                        return -ENOMEM;
+
+                r = safe_atou64(t, &m->since);
+                if (r < 0)
+                        return r;
+
+                m->since_set = true;
         }
 
-        if (!m->cursor)
+        range = colon;
+        colon = strchr(range + 1, ':');
+        if (!colon)
+                until = strdup(range + 1);
+        else {
+                r = request_parse_range_skip_and_n_entries(m, colon);
+                if (r < 0)
+                        return r;
+
+                until = strndup(range + 1, colon - range - 1);
+        }
+
+        if (!until)
                 return -ENOMEM;
 
-        m->cursor[strcspn(m->cursor, WHITESPACE)] = 0;
-        if (isempty(m->cursor))
-                m->cursor = mfree(m->cursor);
+        until[strcspn(until, WHITESPACE)] = 0;
+        if (!isempty(until)) {
+                r = safe_atou64(until, &m->until);
+                if (r < 0) {
+                        free(until);
+                        return r;
+                }
+
+                m->until_set = true;
+        }
+        free(until);
+
+        return 0;
+}
+
+static int request_parse_range(
+                RequestMeta *m,
+                struct MHD_Connection *connection) {
+
+        const char *range;
+
+        assert(m);
+        assert(connection);
+
+        range = MHD_lookup_connection_value(connection, MHD_HEADER_KIND, "Range");
+        if (!range)
+                return 0;
+
+        m->n_skip = 0;
+        if (startswith(range, "entries="))
+                return request_parse_range_entries(m, range);
+        else if (startswith(range, "time="))
```

Comment:
Not a matter, but `else` is redundant.

Reply:
Micro-optimization : when first condition is met, the second is not evaluated (but maybe the compiler handles it automatically ?). If it's more clear without else, I remove it.

Reply:
We usually drop redundant `else`. I hope compiler handles both cases well.

Reply:
Done. And I was wrong : in `if A { return X; } else if B { ... } ` the B is not evaluated when A is true !